File: artifacts/escape-hatch/src/lib/companion-reconciliation.mjs
// Provider-agnostic companion reconciliation (EH-Q1) JS ESM version
// Synthetic only, no provider calls
export const VERIFICATION_STATES = ["SNAPSHOT", "LIVE_VERIFIED", "STALE", "CLOSED", "BLOCKED", "UNKNOWN"];
export const EXECUTION_STATES = ["READY_TO_APPLY", "FILLED", "AWAITING_OPERATOR", "BLOCKED", "SUBMITTED"];
export const CHANNELS = ["web_form", "email", "external_provider"];
export const QUALIFYING_KINDS = new Set(["submission_receipt", "confirmation", "correspondence"]);
export const RECONCILIATION_SCHEMA = "escapehatch/application-companion-reconciliation-result/v1";

function deepClone(v) { return JSON.parse(JSON.stringify(v)); }
function isDateTime(s) {
  if (typeof s !== "string") return false;
  const m = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d+))?(Z|([+-])(\d{2}):(\d{2}))$/.exec(s);
  if (!m) return false;
  const year=Number(m[1]), month=Number(m[2]), day=Number(m[3]);
  const hour=Number(m[4]), minute=Number(m[5]), second=Number(m[6]);
  const offsetHour=m[10]===undefined ? 0 : Number(m[10]);
  const offsetMinute=m[11]===undefined ? 0 : Number(m[11]);
  if (year < 1 || month < 1 || month > 12 || hour > 23 || minute > 59 || second > 59 || offsetHour > 23 || offsetMinute > 59) return false;
  const leap=(year%4===0 && year%100!==0) || year%400===0;
  const monthDays=[31, leap?29:28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  if (day < 1 || day > monthDays[month-1]) return false;
  return !Number.isNaN(Date.parse(s));
}

export function reconcile(careerState, providerSnapshot) {
  const now = new Date().toISOString();
  const reconciled = deepClone(careerState);
  const providerReadBack = Boolean(providerSnapshot && providerSnapshot.read_back === true && isDateTime(providerSnapshot.observed_at || ""));
  const freshness = [];
  const execution = [];
  let anyConflict = false;
  const queueActive = {};
  const historyPreserved = true;

  const oppById = new Map(reconciled.opportunities.map(o => [o.id, o]));
  const providerOppMap = new Map((providerSnapshot?.opportunities || []).map(o => [o.opportunity_id, o]));
  const providerAppMap = new Map((providerSnapshot?.applications || []).map(a => [a.application_id, a]));

  for (const opp of reconciled.opportunities) {
    const from = opp.verification?.state ?? null;
    const providerOpp = providerOppMap.get(opp.id);
    let to = from;
    let active = true;
    let history = true;
    if (!providerReadBack) {
      to = from;
      if (["STALE","CLOSED"].includes(from)) active = false;
      else if (["BLOCKED","UNKNOWN","SNAPSHOT"].includes(from)) active = false;
      else if (from === "LIVE_VERIFIED") active = true;
      else active = false;
    } else if (!providerOpp || !providerOpp.verification) {
      to = from;
      if (["STALE","CLOSED"].includes(from)) active = false;
      else if (from === "LIVE_VERIFIED") active = true;
      else active = false;
    } else {
      const pv = providerOpp.verification;
      if (!VERIFICATION_STATES.includes(pv.state) || !isDateTime(pv.verified_at || "")) {
        to = from;
        active = from === "LIVE_VERIFIED";
      } else if (pv.state === "LIVE_VERIFIED") {
        to = "LIVE_VERIFIED";
        active = true;
        opp.verification = { state: "LIVE_VERIFIED", verified_at: pv.verified_at, detail: pv.detail || "Live posting re-observed via provider read-back", source: pv.source || providerSnapshot.provider_id };
      } else if (["STALE","CLOSED"].includes(pv.state)) {
        to = pv.state;
        active = false;
        opp.verification = { state: pv.state, verified_at: pv.verified_at, detail: pv.detail || `Provider reports ${pv.state}`, source: pv.source || providerSnapshot.provider_id };
      } else if (pv.state === "BLOCKED") {
        to = "BLOCKED";
        active = false;
        opp.verification = { state: "BLOCKED", verified_at: pv.verified_at, detail: pv.detail || "Provider authorization blocked", source: pv.source || providerSnapshot.provider_id };
      } else if (["UNKNOWN","SNAPSHOT"].includes(pv.state)) {
        to = from;
        if (from === "LIVE_VERIFIED" && pv.state === "UNKNOWN") anyConflict = true;
        active = to === "LIVE_VERIFIED";
      }
    }
    queueActive[opp.id] = active;
    freshness.push({ opportunity_id: opp.id, from, to, queue_active: active, history_preserved: history });
  }

  for (const app of reconciled.applications) {
    const from = app.execution?.state ?? null;
    const fromChannel = app.execution?.channel ?? "web_form";
    const oppId = app.opportunity_id;
    const opp = oppById.get(oppId);
    const oppVerification = opp?.verification?.state;
    const providerApp = providerAppMap.get(app.id);
    let to = from;
    let evidenceBinding = "none";
    let conflict = false;
    let reason = "no_change";
    const localQual = reconciled.evidence.filter(e => e.application_id === app.id && QUALIFYING_KINDS.has(e.kind) && isDateTime(e.observed_at || ""));
    const hasLocal = localQual.length > 0;
    const providerQual = (providerApp?.evidence || []).filter(e => QUALIFYING_KINDS.has(e.kind) && isDateTime(e.observed_at || ""));
    const hasProvider = providerQual.length > 0;
    const localAny = reconciled.evidence.filter(e => e.application_id === app.id);
    const hasAnyLocal = localAny.length > 0;
    const providerAny = providerApp?.evidence || [];
    const hasAnyProvider = providerAny.length > 0;

    if (!providerReadBack) {
      to = from; evidenceBinding = hasAnyLocal ? "local" : "none"; reason = "provider_read_back_required";
    } else if (!providerApp) {
      to = from; evidenceBinding = hasAnyLocal ? "local" : "none"; reason = "no_provider_application_snapshot";
      if (["STALE","CLOSED"].includes(oppVerification) && ["READY_TO_APPLY","FILLED"].includes(from)) {
        to = "BLOCKED"; reason = `posting_${oppVerification}_deactivates_queue`;
        if (app.execution) { app.execution.state = "BLOCKED"; app.execution.block_reason = `posting_${oppVerification}_without_erasing_history`; app.execution.last_transition_at = now; }
        else app.execution = { state:"BLOCKED", channel: fromChannel, last_transition_at: now, block_reason:`posting_${oppVerification}` };
      }
    } else {
      const authFailure = providerApp.auth && providerApp.auth.authorized === false;
      const oppAuthFailure = providerOppMap.get(oppId)?.auth?.authorized === false;
      if (authFailure || oppAuthFailure) {
        if (from === "SUBMITTED") {
          conflict = true; anyConflict = true; to = "BLOCKED"; evidenceBinding = hasAnyLocal ? "local" : "none"; reason = "provider_authorization_failure_BLOCKED";
          if (app.execution) { app.execution.state = "BLOCKED"; app.execution.block_reason = providerApp.auth?.reason || "provider_authorization_failure"; app.execution.last_transition_at = now; }
        } else {
          const channel = app.execution?.channel || providerApp.auth?.channel || fromChannel;
          if (channel === "email") {
            to = "AWAITING_OPERATOR"; reason = "provider_authorization_failure_AWAITING_OPERATOR";
            if (app.execution) { app.execution.state = "AWAITING_OPERATOR"; app.execution.awaiting_reason = "provider_authorization_failure"; app.execution.last_transition_at = now; }
          } else {
            to = "BLOCKED"; reason = "provider_authorization_failure_BLOCKED";
            if (app.execution) { app.execution.state = "BLOCKED"; app.execution.block_reason = providerApp.auth?.reason || "provider_authorization_failure"; app.execution.last_transition_at = now; }
          }
          evidenceBinding = hasAnyProvider ? "provider" : hasAnyLocal ? "local" : "none";
        }
      } else {
        const providerFreshness = providerOppMap.get(oppId)?.verification?.state ?? oppVerification;
        if (["STALE","CLOSED"].includes(providerFreshness) && ["READY_TO_APPLY","FILLED"].includes(from)) {
          to = "BLOCKED"; reason = `stale_closed_deactivates_queue_${providerFreshness}`; evidenceBinding = hasAnyLocal ? "local":"none";
          if (app.execution) { app.execution.state = "BLOCKED"; app.execution.block_reason = `posting_${providerFreshness}_deactivates_queue_without_erasing_history`; app.execution.last_transition_at = now; }
        } else if (providerFreshness === "BLOCKED" && from === "SUBMITTED") {
          to = "BLOCKED"; reason = "verification_BLOCKED_blocks_SUBMITTED"; conflict=true; anyConflict=true; evidenceBinding = hasAnyProvider ? "provider" : hasAnyLocal ? "local":"none";
          if (app.execution) { app.execution.state="BLOCKED"; app.execution.block_reason="verification_BLOCKED"; app.execution.last_transition_at=now; }
        } else {
          const channel = app.execution?.channel || providerApp.execution?.channel || "web_form";
          const wantsSubmitted = (providerApp.execution?.state === "SUBMITTED") || from === "SUBMITTED";
          const isEmail = channel === "email";
          let emailGuardPassed = true;
          if (isEmail && wantsSubmitted) {
            const mail = providerApp.mail || {};
            const operatorConfirmed = providerApp.operator_confirmed === true;
            const hasSent = mail.sent === true || mail.provider_confirmation === "sent" || mail.outbox === true;
            if (!hasSent && !operatorConfirmed) {
              emailGuardPassed = false;
              if (from === "SUBMITTED") {
                conflict=true; anyConflict=true; to="AWAITING_OPERATOR"; reason="draft_email_requires_sent_confirmation"; evidenceBinding=hasAnyLocal?"local":"none";
                if (app.execution) { app.execution.state="AWAITING_OPERATOR"; app.execution.awaiting_reason="draft_not_sent_requires_provider_confirmation_or_operator"; app.execution.last_transition_at=now; }
              } else {
                to = from==="FILLED" ? "FILLED":"AWAITING_OPERATOR"; reason="email_draft_not_sent_requires_confirmation"; evidenceBinding=hasAnyProvider?"provider":hasAnyLocal?"local":"none";
              }
            }
          }
          if (emailGuardPassed) {
            if (wantsSubmitted) {
              const hasQualifying = hasProvider || hasLocal || providerApp.operator_confirmed === true;
              if (!hasQualifying) {
                if (from==="FILLED") to="FILLED"; else if (from==="READY_TO_APPLY") to="READY_TO_APPLY"; else to="AWAITING_OPERATOR";
                reason="SUBMITTED_requires_qualifying_evidence"; evidenceBinding="none"; if (hasLocal && !hasProvider){conflict=true; anyConflict=true;}
              } else {
                const providerStronger = hasProvider && (!hasLocal || providerQual.length >= localQual.length);
                const localStronger = hasLocal && !hasProvider;
                if (providerStronger) {
                  to="SUBMITTED"; evidenceBinding=providerApp.operator_confirmed ? "operator_confirmed":"provider"; reason="provider_qualifying_evidence_promotes_SUBMITTED";
                  if (hasLocal && providerQual[0]?.id !== localQual[0]?.id){conflict=true; anyConflict=true;}
                  if (app.execution) {
                    app.execution.state="SUBMITTED"; app.execution.evidence_id=providerQual[0]?.id || app.execution.evidence_id || localQual[0]?.id; app.execution.last_transition_at=providerApp.execution?.last_transition_at || now;
                    if (!app.submitted_at) app.submitted_at=now; if (!app.external_reference) app.external_reference=`provider-${providerSnapshot.provider_id}-${app.id}`;
                  } else {
                    app.execution={state:"SUBMITTED", channel, last_transition_at:providerApp.execution?.last_transition_at || now, evidence_id: providerQual[0]?.id || localQual[0]?.id};
                    if (!app.submitted_at) app.submitted_at=now;
                  }
                  for (const pev of providerQual){ if (!reconciled.evidence.some(e=>e.id===pev.id)) reconciled.evidence.push({id:pev.id, application_id:app.id, kind:pev.kind, artifact:{owner:"user",kind:"relative_path",locator:`evidence/${app.id}/${pev.id}.txt`}, observed_at:pev.observed_at||now}); }
                } else if (localStronger) {
                  to = from==="SUBMITTED" ? "SUBMITTED":"FILLED"; evidenceBinding="local"; reason="local_qualifying_preserved_conflict"; if (hasProvider){conflict=true; anyConflict=true;}
                } else {
                  to="SUBMITTED"; evidenceBinding=hasProvider?"provider":"local"; reason="conflict_preserved_keep_stronger"; conflict=true; anyConflict=true;
                }
              }
            } else {
              to=from;
              if (hasAnyLocal && !hasAnyProvider){evidenceBinding="local"; reason="local_export_remains_local_until_reconciled";}
              else if (hasAnyProvider){evidenceBinding="provider"; reason="provider_evidence_available"; if (hasLocal && providerQual[0]?.id !== localQual[0]?.id){conflict=true; anyConflict=true; reason="conflict_preserved_provider_stronger";} else if (hasAnyLocal && providerAny[0]?.id !== localAny[0]?.id){conflict=true; anyConflict=true;}}
              else {evidenceBinding="none"; reason="no_qualifying_evidence";}
              if (["STALE","CLOSED"].includes(oppVerification) && ["READY_TO_APPLY","FILLED"].includes(from)){
                to="BLOCKED"; reason=`freshness_${oppVerification}_deactivates_queue`;
                if (app.execution){app.execution.state="BLOCKED"; app.execution.block_reason=`posting_${oppVerification}`; app.execution.last_transition_at=now;}
              }
            }
          }
        }
      }
    }
    if (providerApp?.operator_confirmed) evidenceBinding="operator_confirmed";
    execution.push({application_id:app.id, from, to, evidence_binding: evidenceBinding, conflict_preserved: conflict, reason});
    if (to!==from && app.execution && app.execution.state!==to) app.execution.state=to;
    if (conflict) anyConflict=true;
  }

  let overall="none";
  if (execution.some(e=>e.evidence_binding==="operator_confirmed")) overall="operator_confirmed";
  else if (execution.some(e=>e.evidence_binding==="provider")) overall="provider";
  else if (execution.some(e=>e.evidence_binding==="local")) overall="local";

  const result={schema:RECONCILIATION_SCHEMA, reconciled_at:now, provider_read_back:providerReadBack, freshness_transition:freshness, execution_transition:execution, evidence_binding:overall, conflict_preserved:anyConflict, queue_active:queueActive, history_preserved:historyPreserved, idempotent:true};
  if (JSON.stringify(careerState)!==JSON.stringify(reconciled)){
    reconciled.revision=(reconciled.revision||0)+1;
    reconciled.updated_at=now;
  }
  return {reconciledState:reconciled, result};
}

export function isIdempotent(careerState, snapshot){
  const first=reconcile(careerState, snapshot);
  const second=reconcile(first.reconciledState, snapshot);
  return JSON.stringify(first.reconciledState)===JSON.stringify(second.reconciledState);
}
export function requiresProviderReadBack(snapshot){ return !snapshot || snapshot.read_back!==true || !isDateTime(snapshot.observed_at || ""); }

File: scripts/companion_reconciliation.py
#!/usr/bin/env python3
"""Provider-agnostic companion reconciliation (EH-Q1). Synthetic only, no provider calls."""

from __future__ import annotations

import copy
import json
import re
from datetime import datetime, timezone
from typing import Any, Dict, List, Tuple

VERIFICATION_STATES = {"SNAPSHOT", "LIVE_VERIFIED", "STALE", "CLOSED", "BLOCKED", "UNKNOWN"}
EXECUTION_STATES = {"READY_TO_APPLY", "FILLED", "AWAITING_OPERATOR", "BLOCKED", "SUBMITTED"}
CHANNELS = {"web_form", "email", "external_provider"}
QUALIFYING_KINDS = {"submission_receipt", "confirmation", "correspondence"}
RECONCILIATION_SCHEMA = "escapehatch/application-companion-reconciliation-result/v1"


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


_RFC3339_DATE_TIME = re.compile(
    r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})$"
)

def _is_datetime(s: str) -> bool:
    if not isinstance(s, str) or not _RFC3339_DATE_TIME.fullmatch(s):
        return False
    try:
        v = s[:-1] + "+00:00" if s.endswith("Z") else s
        datetime.fromisoformat(v)
        return True
    except (TypeError, ValueError):
        return False


def reconcile(career_state: Dict[str, Any], provider_snapshot: Dict[str, Any] | None) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    now = _now_iso()
    reconciled = copy.deepcopy(career_state)
    provider_read_back = bool(provider_snapshot and provider_snapshot.get("read_back") is True and _is_datetime(provider_snapshot.get("observed_at", "")))

    freshness: List[Dict[str, Any]] = []
    execution: List[Dict[str, Any]] = []
    any_conflict = False
    queue_active: Dict[str, bool] = {}
    history_preserved = True

    opp_by_id = {o["id"]: o for o in reconciled.get("opportunities", [])}
    app_by_id = {a["id"]: a for a in reconciled.get("applications", [])}
    ev_by_app: Dict[str, List[Dict[str, Any]]] = {}
    for ev in reconciled.get("evidence", []):
        ev_by_app.setdefault(ev["application_id"], []).append(ev)

    provider_opp_map = {o["opportunity_id"]: o for o in (provider_snapshot or {}).get("opportunities", [])}
    provider_app_map = {a["application_id"]: a for a in (provider_snapshot or {}).get("applications", [])}

    # Freshness per opportunity
    for opp in reconciled.get("opportunities", []):
        opp_id = opp["id"]
        from_state = opp.get("verification", {}).get("state") if opp.get("verification") else None
        provider_opp = provider_opp_map.get(opp_id)
        to_state = from_state
        active = True
        history = True

        if not provider_read_back:
            to_state = from_state
            if from_state in ("STALE", "CLOSED"):
                active = False
            elif from_state in ("BLOCKED", "UNKNOWN", "SNAPSHOT"):
                active = False
            elif from_state == "LIVE_VERIFIED":
                active = True
            else:
                active = False if from_state else False
        elif not provider_opp or not provider_opp.get("verification"):
            to_state = from_state
            if from_state in ("STALE", "CLOSED"):
                active = False
            elif from_state == "LIVE_VERIFIED":
                active = True
            else:
                active = False
        else:
            pv = provider_opp["verification"]
            p_state = pv.get("state")
            p_verified_at = pv.get("verified_at")
            if p_state not in VERIFICATION_STATES or not _is_datetime(p_verified_at or ""):
                to_state = from_state
                active = from_state == "LIVE_VERIFIED"
            elif p_state == "LIVE_VERIFIED":
                to_state = "LIVE_VERIFIED"
                active = True
                opp["verification"] = {
                    "state": "LIVE_VERIFIED",
                    "verified_at": p_verified_at,
                    "detail": pv.get("detail") or "Live posting re-observed via provider read-back",
                    "source": pv.get("source") or (provider_snapshot or {}).get("provider_id", "synthetic"),
                }
            elif p_state in ("STALE", "CLOSED"):
                to_state = p_state
                active = False
                opp["verification"] = {
                    "state": p_state,
                    "verified_at": p_verified_at,
                    "detail": pv.get("detail") or f"Provider reports {p_state}",
                    "source": pv.get("source") or (provider_snapshot or {}).get("provider_id", "synthetic"),
                }
                history = True
            elif p_state == "BLOCKED":
                to_state = "BLOCKED"
                active = False
                opp["verification"] = {
                    "state": "BLOCKED",
                    "verified_at": p_verified_at,
                    "detail": pv.get("detail") or "Provider authorization blocked",
                    "source": pv.get("source") or (provider_snapshot or {}).get("provider_id", "synthetic"),
                }
            elif p_state in ("UNKNOWN", "SNAPSHOT"):
                to_state = from_state
                if from_state == "LIVE_VERIFIED" and p_state == "UNKNOWN":
                    any_conflict = True
                active = to_state == "LIVE_VERIFIED"

        queue_active[opp_id] = active
        freshness.append({
            "opportunity_id": opp_id,
            "from": from_state,
            "to": to_state,
            "queue_active": active,
            "history_preserved": history,
        })

    # Execution per application
    for app in reconciled.get("applications", []):
        app_id = app["id"]
        opp_id = app["opportunity_id"]
        from_state = app.get("execution", {}).get("state") if app.get("execution") else None
        from_channel = app.get("execution", {}).get("channel") if app.get("execution") else "web_form"
        opp = opp_by_id.get(opp_id)
        opp_verification = opp.get("verification", {}).get("state") if opp and opp.get("verification") else None
        provider_app = provider_app_map.get(app_id)

        to_state = from_state
        evidence_binding: str = "none"
        conflict = False
        reason = "no_change"

        local_qual = [e for e in reconciled.get("evidence", []) if e["application_id"] == app_id and e["kind"] in QUALIFYING_KINDS and _is_datetime(e.get("observed_at", ""))]
        has_local = len(local_qual) > 0
        provider_qual = [e for e in (provider_app or {}).get("evidence", []) if e.get("kind") in QUALIFYING_KINDS and _is_datetime(e.get("observed_at", ""))]
        has_provider = len(provider_qual) > 0
        # any evidence (even note/draft) counts for local binding
        local_any = [e for e in reconciled.get("evidence", []) if e["application_id"] == app_id]
        has_any_local = len(local_any) > 0
        provider_any = (provider_app or {}).get("evidence", []) or []
        has_any_provider = len(provider_any) > 0

        if not provider_read_back:
            to_state = from_state
            evidence_binding = "local" if has_any_local else "none"
            reason = "provider_read_back_required"
        elif not provider_app:
            to_state = from_state
            evidence_binding = "local" if has_any_local else "none"
            reason = "no_provider_application_snapshot"
            if opp_verification in ("STALE", "CLOSED") and from_state in ("READY_TO_APPLY", "FILLED"):
                to_state = "BLOCKED"
                reason = f"posting_{opp_verification}_deactivates_queue"
                if app.get("execution"):
                    app["execution"]["state"] = "BLOCKED"
                    app["execution"]["block_reason"] = f"posting_{opp_verification}_without_erasing_history"
                    app["execution"]["last_transition_at"] = now
                else:
                    app["execution"] = {"state": "BLOCKED", "channel": from_channel, "last_transition_at": now, "block_reason": f"posting_{opp_verification}"}
        else:
            auth_failure = provider_app.get("auth") and provider_app["auth"].get("authorized") is False
            opp_auth_failure = provider_opp_map.get(opp_id, {}).get("auth", {}).get("authorized") is False if provider_opp_map.get(opp_id, {}).get("auth") else False
            if auth_failure or opp_auth_failure:
                if from_state == "SUBMITTED":
                    conflict = True
                    any_conflict = True
                    to_state = "BLOCKED"
                    evidence_binding = "local" if has_local else "none"
                    reason = "provider_authorization_failure_BLOCKED"
                    if app.get("execution"):
                        app["execution"]["state"] = "BLOCKED"
                        app["execution"]["block_reason"] = provider_app.get("auth", {}).get("reason") or "provider_authorization_failure"
                        app["execution"]["last_transition_at"] = now
                else:
                    channel = app.get("execution", {}).get("channel") or provider_app.get("auth", {}).get("channel") or from_channel
                    if channel == "email":
                        to_state = "AWAITING_OPERATOR"
                        reason = "provider_authorization_failure_AWAITING_OPERATOR"
                        if app.get("execution"):
                            app["execution"]["state"] = "AWAITING_OPERATOR"
                            app["execution"]["awaiting_reason"] = "provider_authorization_failure"
                            app["execution"]["last_transition_at"] = now
                    else:
                        to_state = "BLOCKED"
                        reason = "provider_authorization_failure_BLOCKED"
                        if app.get("execution"):
                            app["execution"]["state"] = "BLOCKED"
                            app["execution"]["block_reason"] = provider_app.get("auth", {}).get("reason") or "provider_authorization_failure"
                            app["execution"]["last_transition_at"] = now
                    evidence_binding = "provider" if has_any_provider else "local" if has_any_local else "none"
            else:
                provider_freshness = provider_opp_map.get(opp_id, {}).get("verification", {}).get("state") if provider_opp_map.get(opp_id, {}).get("verification") else opp_verification
                if provider_freshness in ("STALE", "CLOSED") and from_state in ("READY_TO_APPLY", "FILLED"):
                    to_state = "BLOCKED"
                    reason = f"stale_closed_deactivates_queue_{provider_freshness}"
                    evidence_binding = "local" if has_any_local else "none"
                    if app.get("execution"):
                        app["execution"]["state"] = "BLOCKED"
                        app["execution"]["block_reason"] = f"posting_{provider_freshness}_deactivates_queue_without_erasing_history"
                        app["execution"]["last_transition_at"] = now
                elif provider_freshness == "BLOCKED" and from_state == "SUBMITTED":
                    to_state = "BLOCKED"
                    reason = "verification_BLOCKED_blocks_SUBMITTED"
                    conflict = True
                    any_conflict = True
                    evidence_binding = "provider" if has_any_provider else "local" if has_any_local else "none"
                    if app.get("execution"):
                        app["execution"]["state"] = "BLOCKED"
                        app["execution"]["block_reason"] = "verification_BLOCKED"
                        app["execution"]["last_transition_at"] = now
                else:
                    channel = (app.get("execution", {}) or {}).get("channel") or provider_app.get("execution", {}).get("channel") or "web_form"
                    wants_submitted = (provider_app.get("execution", {}) or {}).get("state") == "SUBMITTED" or from_state == "SUBMITTED"
                    is_email = channel == "email"
                    email_guard_passed = True
                    if is_email and wants_submitted:
                        mail = provider_app.get("mail") or {}
                        operator_confirmed = provider_app.get("operator_confirmed") is True
                        has_sent = mail.get("sent") is True or mail.get("provider_confirmation") == "sent" or mail.get("outbox") is True
                        if not has_sent and not operator_confirmed:
                            email_guard_passed = False
                            if from_state == "SUBMITTED":
                                conflict = True
                                any_conflict = True
                                to_state = "AWAITING_OPERATOR"
                                reason = "draft_email_requires_sent_confirmation"
                                evidence_binding = "local" if has_any_local else "none"
                                if app.get("execution"):
                                    app["execution"]["state"] = "AWAITING_OPERATOR"
                                    app["execution"]["awaiting_reason"] = "draft_not_sent_requires_provider_confirmation_or_operator"
                                    app["execution"]["last_transition_at"] = now
                            else:
                                to_state = "FILLED" if from_state == "FILLED" else "AWAITING_OPERATOR"
                                reason = "email_draft_not_sent_requires_confirmation"
                                evidence_binding = "provider" if has_any_provider else "local" if has_any_local else "none"
                    if email_guard_passed:
                        if wants_submitted:
                            has_qualifying = has_provider or has_local or provider_app.get("operator_confirmed") is True
                            if not has_qualifying:
                                if from_state == "FILLED":
                                    to_state = "FILLED"
                                elif from_state == "READY_TO_APPLY":
                                    to_state = "READY_TO_APPLY"
                                else:
                                    to_state = "AWAITING_OPERATOR"
                                reason = "SUBMITTED_requires_qualifying_evidence"
                                evidence_binding = "none"
                                if has_local and not has_provider:
                                    conflict = True
                                    any_conflict = True
                            else:
                                provider_stronger = has_provider and (not has_local or len(provider_qual) >= len(local_qual))
                                local_stronger = has_local and not has_provider
                                if provider_stronger:
                                    to_state = "SUBMITTED"
                                    evidence_binding = "operator_confirmed" if provider_app.get("operator_confirmed") else "provider"
                                    reason = "provider_qualifying_evidence_promotes_SUBMITTED"
                                    if has_local and provider_qual and local_qual and provider_qual[0].get("id") != local_qual[0].get("id"):
                                        conflict = True
                                        any_conflict = True
                                    if app.get("execution"):
                                        app["execution"]["state"] = "SUBMITTED"
                                        app["execution"]["evidence_id"] = provider_qual[0].get("id") if provider_qual else app["execution"].get("evidence_id") or local_qual[0].get("id")
                                        app["execution"]["last_transition_at"] = provider_app.get("execution", {}).get("last_transition_at") or now
                                        if not app.get("submitted_at"):
                                            app["submitted_at"] = now
                                        if not app.get("external_reference"):
                                            app["external_reference"] = f"provider-{provider_snapshot.get('provider_id','synthetic')}-{app_id}"
                                    else:
                                        app["execution"] = {"state": "SUBMITTED", "channel": channel, "last_transition_at": provider_app.get("execution", {}).get("last_transition_at") or now, "evidence_id": provider_qual[0].get("id") if provider_qual else local_qual[0].get("id")}
                                        if not app.get("submitted_at"):
                                            app["submitted_at"] = now
                                    for pev in provider_qual:
                                        if not any(e["id"] == pev.get("id") for e in reconciled.get("evidence", [])):
                                            reconciled["evidence"].append({"id": pev.get("id"), "application_id": app_id, "kind": pev.get("kind"), "artifact": {"owner": "user", "kind": "relative_path", "locator": f"evidence/{app_id}/{pev.get('id')}.txt"}, "observed_at": pev.get("observed_at") or now})
                                elif local_stronger:
                                    to_state = "SUBMITTED" if from_state == "SUBMITTED" else "FILLED"
                                    evidence_binding = "local"
                                    reason = "local_qualifying_preserved_conflict"
                                    if has_provider:
                                        conflict = True
                                        any_conflict = True
                                else:
                                    to_state = "SUBMITTED"
                                    evidence_binding = "provider" if has_provider else "local"
                                    reason = "conflict_preserved_keep_stronger"
                                    conflict = True
                                    any_conflict = True
                        else:
                            to_state = from_state
                            if has_any_local and not has_any_provider:
                                evidence_binding = "local"
                                reason = "local_export_remains_local_until_reconciled"
                            elif has_any_provider:
                                evidence_binding = "provider"
                                reason = "provider_evidence_available"
                                if has_local and provider_qual and local_qual and provider_qual[0].get("id") != local_qual[0].get("id"):
                                    conflict = True
                                    any_conflict = True
                                    reason = "conflict_preserved_provider_stronger"
                            else:
                                evidence_binding = "none"
                                reason = "no_qualifying_evidence"
                            if opp_verification in ("STALE", "CLOSED") and from_state in ("READY_TO_APPLY", "FILLED"):
                                to_state = "BLOCKED"
                                reason = f"freshness_{opp_verification}_deactivates_queue"
                                if app.get("execution"):
                                    app["execution"]["state"] = "BLOCKED"
                                    app["execution"]["block_reason"] = f"posting_{opp_verification}"
                                    app["execution"]["last_transition_at"] = now

        if provider_app and provider_app.get("operator_confirmed"):
            evidence_binding = "operator_confirmed"

        execution.append({
            "application_id": app_id,
            "from": from_state,
            "to": to_state,
            "evidence_binding": evidence_binding,
            "conflict_preserved": conflict,
            "reason": reason,
        })
        if to_state != from_state and app.get("execution"):
            if app["execution"]["state"] != to_state:
                app["execution"]["state"] = to_state
        if conflict:
            any_conflict = True

    # overall evidence_binding
    overall = "none"
    if any(e["evidence_binding"] == "operator_confirmed" for e in execution):
        overall = "operator_confirmed"
    elif any(e["evidence_binding"] == "provider" for e in execution):
        overall = "provider"
    elif any(e["evidence_binding"] == "local" for e in execution):
        overall = "local"

    result = {
        "schema": RECONCILIATION_SCHEMA,
        "reconciled_at": now,
        "provider_read_back": provider_read_back,
        "freshness_transition": freshness,
        "execution_transition": execution,
        "evidence_binding": overall,
        "conflict_preserved": any_conflict,
        "queue_active": queue_active,
        "history_preserved": history_preserved,
        "idempotent": True,
    }

    # Any persisted mutation (including evidence-only reconciliation) advances revision.
    if json.dumps(career_state, sort_keys=True) != json.dumps(reconciled, sort_keys=True):
        reconciled["revision"] = (reconciled.get("revision") or 0) + 1
        reconciled["updated_at"] = now

    return reconciled, result


def is_idempotent(career_state: Dict[str, Any], snapshot: Dict[str, Any]) -> bool:
    first_state, first_result = reconcile(career_state, snapshot)
    second_state, second_result = reconcile(first_state, snapshot)
    return json.dumps(first_state, sort_keys=True) == json.dumps(second_state, sort_keys=True)


def requires_provider_read_back(snapshot: Dict[str, Any] | None) -> bool:
    return not snapshot or snapshot.get("read_back") is not True or not _is_datetime(snapshot.get("observed_at", ""))
